### Brief Overview 

- The goal of this assignment is to practice some important data wrangling functionality commonly required in real-world projects.

- Here we will use two datasets:
  - IRS Statistics of Income (SOI) dataset
  - The Medicaid Data per State  


- The final product here is a table with medication cost per Medicaid enrollee per state. This dataset will allow us to answer such questions as:
  - Which medications account for the bulk of a state's spending   
  - Which drugs are prescribed much more in one state compared to the other states.
etc.


In [1]:
# import libraries
import pandas as pd
import numpy as np


* Load the IRS Statistics of Income (SOI) dataset (tax_data.csv) into a `DataFrame` called `tax_data`. The file is `tax_data.csv` is located in the `data` directory of the assignment folder.

* This dataset was preprocessed but the original one was obtained at the following URL:

        https://www.irs.gov/pub/irs-soi/15zpallagi.csv


In [2]:
tax_data = pd.read_csv("data/tax_data.csv")

* Use a `tax_data` method or property to display the first eight (8) rows of the `DataFrame`

In [3]:
tax_data.head(8)

STATEFIPS STATE  zipcode  agi_stub        N1     mars1     MARS2     MARS4  \
0        1.0    AL      0.0       1.0  836320.0  481570.0  109790.0  233260.0   
1        1.0    AL      0.0       2.0  494830.0  206630.0  146250.0  129390.0   
2        1.0    AL      0.0       3.0  261250.0   80720.0  139280.0   36130.0   
3        1.0    AL      0.0       4.0  166690.0   28510.0  124650.0   10630.0   
4        1.0    AL      0.0       5.0  212660.0   19520.0  184320.0    4830.0   
5        1.0    AL      0.0       6.0   55360.0    2950.0   49260.0     350.0   
6        1.0    AL  35004.0       1.0    1490.0     970.0     230.0     280.0   
7        1.0    AL  35004.0       2.0    1350.0     630.0     360.0     300.0   

       PREP         N2  ...    N10300     A10300   N85530   A85530   N85300  \
0  455560.0  1356760.0  ...  373410.0   328469.0      0.0      0.0      0.0   
1  275920.0  1010990.0  ...  395880.0   965011.0      0.0      0.0      0.0   
2  155100.0   583910.0  ...  251490.0  1333418.0      0.0      0.0      0.0   
3   99950.0   423990.0  ...  165320.0  1414283.0      0.0      0.0      0.0   
4  126860.0   589490.0  ...  212000.0  3820152.0    420.0    168.0     60.0   
5   41410.0   160530.0  ...   55300.0  6027793.0  22090.0  39519.0  27550.0   
6     700.0     2160.0  ...     690.0      610.0      0.0      0.0      0.0   
7     610.0     2540.0  ...    1140.0     3019.0      0.0      0.0      0.0   

    A85300   N11901    A11901    N11902     A11902  
0      0.0  61920.0   48150.0  732670.0  1933120.0  
1      0.0  73720.0  107304.0  415410.0  1187403.0  
2      0.0  64200.0  139598.0  193030.0   536699.0  
3      0.0  45460.0  128823.0  116440.0   377177.0  
4     31.0  83330.0  421004.0  121570.0   483682.0  
5  95112.0  28590.0  791573.0   15960.0   250289.0  
6      0.0    120.0      94.0    1290.0     2792.0  
7      0.0    210.0     301.0    1130.0     2935.0  

[8 rows x 91 columns]

*  Modify `tax_data` to uppercase all the header name. 
  * Your answer can only use `DataFrame` or `Series` methods or properties
  * Do not hardcode the operation by uppercasing the columns yourself
  *  You can `tax_data.columns`, which returns a `Series` of the column names.
  
* The resulting column name should look as follows:


```
STATEFIPS    STATE    ZIPCODE    AGI_STUB    N1    MARS1    MARS2    MARS4    PREP    N2    ...    
```

This operation is useful for standardizing column names and avoid guessing whether the column header was in upper case, lower case or a mix of both.

In [4]:
tax_data.columns = tax_data.columns.str.upper()

* What is the total number of entries (also called observations) in `tax_data`?

  * Your answer can only use `DataFrame` or `Series` methods or properties

In [5]:
tax_data.shape

(166698, 91)

- If `STATEFIPS` is header title of the first column of the `tax_data` `DataFrame`, what is the title of the 32nd column
  - Your answer can only use `DataFrame` or `Series` methods or properties and should use a single python expression



In [6]:
tax_data.columns[31]

'N00900'

* If `STATEFIPS` is the the first column, what is the index of the column name `N10300`?

  * Your answer can only use `DataFrame` or `Series` methods or properties


In [7]:
tax_data.columns.get_loc("N10300")

81

* Print the count of unique zip codes in each state using descending order. Your answer can only use `DataFrame` or `Series` methods or properties. The result should look like the following ( '...' represents remaining data that is not shown),
    ```python
        ZIPCODE
    STATE
    TX          9714
    NY          9238
    CA          8908
    PA          8208
    IL          7386
    ...
    ```
  The above indicates that there are 9,714 unique zipcodes in TX, 9,238 in NY, etc."

In [8]:
uniq_zip = tax_data.groupby('STATE')['ZIPCODE'].count().sort_values(ascending=False)
uniq_zip

STATE
TX    9714
NY    9238
CA    8908
PA    8208
IL    7386
OH    5988
FL    5520
MI    5352
MO    5340
IA    4962
VA    4758
MN    4751
NC    4350
WI    4284
IN    4044
GA    4002
KY    3936
KS    3588
TN    3540
AL    3462
NJ    3282
OK    3282
WV    3072
WA    2982
AR    2940
NE    2916
MA    2892
LA    2718
MD    2430
CO    2370
SC    2256
MS    2220
ME    2220
OR    2136
AZ    1770
ND    1722
SD    1722
CT    1584
VT    1428
NH    1392
MT    1350
ID    1278
NM    1254
UT    1116
NV     768
WY     654
RI     426
HI     360
DE     342
AK     336
DC     144
Name: ZIPCODE, dtype: int64

- Identify the position of HI in the list of zip code counts per state (questions directly above)
  - Your answer can only use `DataFrame` or `Series` methods or properties

In [9]:
df = pd.DataFrame(uniq_zip)
df.index.get_loc('HI')

47

### Identifying and Removing Ambiguous Zip Codes

- Count the number of entries where ZIPCODE is 0, assign your results to a variable named  `nb_invalid_zip`

In [10]:
nb_invalid_zip = int((tax_data['ZIPCODE'] == 0).sum())
nb_invalid_zip

306

* Run the line below to make sure that `nb_invalid_zip` is an integer (`int`)
  * Note that `assert` will only print an error if `type(nb_invalid_zip)` is not of type `int`
  
* If `nb_invalid_zip` then change your answer above so that the value returned is effectively a number.

In [11]:
assert(type(nb_invalid_zip) == int)

- Remove from `tax_data` all the lines where the zip code is `0` and save resulting `DataFrame` to a variable named `tax_data_valid_zip`
  - Your answer can only use `DataFrame` or `Series` methods or properties


In [12]:
tax_data_valid_zip = tax_data[(tax_data[['ZIPCODE']] != 0).all(axis=1)]

* Run the line below to confirm that the operation worked as expected

  * The assertion below is testing that the number of lines with "zip code equal to  0" + number of lines in `tax_data_valid_zip` is equal to the number of lines in the original `DataFrame` `tax_data`
  
  * The assertion below will fail (and print an error message) if the results do not match. If that is the case, please review your code above.

In [13]:
assert((tax_data_valid_zip.shape[0] + nb_invalid_zip) == tax_data.shape[0])

### Identifying and Removing Lines with Missing Values

* How many lines contain at least one missing value `NaN` in the `tax_data_valid_zip` `DataFrame`?
  * Your answer can only use `DataFrame` methods and properties
* Assing the count of `NaN` into a variable called nb_missing_values

In [14]:
nb_missing_values = tax_data_valid_zip.isna().sum().sum()
nb_missing_values

139

* Create a new `DataFrame` containing all the lines from `tax_data_valid_zip`, except lines containing missing values

* Call the new `DataFrame` `tax_data_valid_zip_cleaned

In [15]:
tax_data_valid_zip_cleaned = tax_data_valid_zip.dropna()
tax_data_valid_zip_cleaned.shape

(166253, 91)

* Run the line below to confirm that the operation worked as expected. The assertion below is testing that:  
`nb_missing_values` + number of lines in `tax_data_valid_zip_cleaned` is equal to the number of lines in `tax_data_valid_zip`
  
* Note that assert will only print an error if the results do not match

In [16]:
assert((tax_data_valid_zip_cleaned.shape[0] + nb_missing_values) == tax_data_valid_zip.shape[0])

### Computing the Percentile Income per Zip Code

* The function `compute_percentile_zipcode` below computes the percentile income per zip code

* By default percentile=0.5,  i.e., the function computes the median

* Read the code and make sure you understand what it does before moving on to the next question


In [17]:
def compute_percentile_zip(df_zip, percentile=0.5):
    index_median = sum(( df_zip["N1"]/ sum(df_zip["N1"])).cumsum() <= percentile)
    val_below_or_at_median = (df_zip["A00100"] /df_zip["N1"]).iloc[index_median]
    return val_below_or_at_median

* Compute the 65th income percentile ( percentile=0.65) for each zipcode in `tax_data_valid_zip_cleaned` data frame

  * Your answer can only use `DataFrame` or `Series` methods or properties
  
  * Recall that you can use the `apply` function to transform the values of a column
  
  
* Sort the values in descending order and assign them to a new `DataFrame` called `zip_rev_all`
* The resulting `Series` should look like the following( '...' represents remaining data that is not shown)

```Python
ZIPCODE
33109    3954.114286
33480    3413.301538
...
```

In [18]:
zip_rev_all = (tax_data.groupby('ZIPCODE').apply(compute_percentile_zip, percentile=0.65)).sort_values(ascending=False)
zip_rev_all

ZIPCODE
33109.0    3954.114286
33480.0    3413.301538
94301.0    3109.443711
94027.0    3091.537013
10577.0    2414.855556
              ...     
53706.0       7.292857
55455.0       7.210000
48109.0       6.981250
84112.0       6.597297
59053.0       5.421429
Length: 27683, dtype: float64

- What are the three zip codes with the most significant 65th percentile value for income?

In [19]:
zip_rev_all.head(3)

ZIPCODE
33109.0    3954.114286
33480.0    3413.301538
94301.0    3109.443711
dtype: float64

# 2 Working with the Medicaid Data

### Loading and exploring the data 

* Load the Medicaid data stored in the file `medicaid_data.csv` into a `DataFrame` called `medicaid_data`. The file is located in the `data` directory of the assignment folder. 
* Note that this is quite large and may take some time to load on a computer with modest RAM resources (4GB or less)


In [20]:
medicaid_data = pd.read_csv("data/medicaid_data.csv")
medicaid_data.head()

Utilization Type State          NDC Product Name  Units Reimbursed  \
0             MCOU    PA  55150023930   Dexamethas              33.0   
1             FFSU    NY     23917710   ALPHAGAN P             570.0   
2             MCOU    OR  13925050501   Dapsone 10             456.0   
3             FFSU    MN  51862006401     DIAZEPAM             780.0   
4             FFSU    MN    781237101   DEXTROAMPH             451.0   

   Number of Prescriptions  Total Amount Reimbursed  \
0                     19.0                   234.98   
1                     57.0                 16006.34   
2                     15.0                  1052.42   
3                     16.0                    89.60   
4                     12.0                  1411.24   

   Medicaid Amount Reimbursed  Non Medicaid Amount Reimbursed  \
0                      234.98                            0.00   
1                    16006.34                            0.00   
2                     1052.42                            0.00   
3                       77.60                           12.00   
4                      198.93                         1212.31   

               Location  
0    (40.5773, -77.264)  
1   (42.1497, -74.9384)  
2  (44.5672, -122.1269)  
3   (45.7326, -93.9196)  
4   (45.7326, -93.9196)


- Modify `medicaid_data` to uppercase all the column names 

  - If your solution uses an assignment, the righthand side of the assignment (rvalue) can only use `DataFrame` or `Series` methods or properties


In [21]:
medicaid_data.columns = medicaid_data.columns.str.upper()


- Familiarize your self with the data
  - the `NDC` column stands for National Drug Code, a universal product identifier for human drugs in the United States
  
  - The remaining column names are self-explanatory
  
- Explore the number of lines and columns in the data

- Check that your column headers are in uppercase

In [22]:
medicaid_data.shape

(1695546, 10)

In [23]:
medicaid_data.columns

Index(['UTILIZATION TYPE', 'STATE', 'NDC', 'PRODUCT NAME', 'UNITS REIMBURSED',
       'NUMBER OF PRESCRIPTIONS', 'TOTAL AMOUNT REIMBURSED',
       'MEDICAID AMOUNT REIMBURSED', 'NON MEDICAID AMOUNT REIMBURSED',
       'LOCATION'],
      dtype='object')

* If you explore the  `Location` column for all the entries for which "STATE" value is equal to "HI" you'll notice that all the values are identical

* Are there any states that have more than one value for `Location`. 
  * Hint: think about using a sorted aggregation as part of a split-apply-combine operation to answer this question
  - Your answer can only use `DataFrame` or `Series` methods or properties


In [24]:
state_locations = medicaid_data.groupby('STATE')['LOCATION'].agg(set)
states_with_multiple_locations = state_locations[state_locations.apply(len) > 1]
print(states_with_multiple_locations)

Series([], Name: LOCATION, dtype: object)




* To compare medication prescriptions across states in a fair and balanced way, we need the number of Medicaid beneficiaries in each state. The following example illustrates the importance of normalizing the values `UNITS REIMBURSED` for each medication in each state by the number of Medicaid enrollees in each state.
  
* The `medicaid_data` DataFrame shows that for the drug with NDC `61958180101` (the drug name is HARVONI and it's used to treat Hepatitis C) there were 11,886  units sold in KY, versus 40,142 in CA -- that's almost 4 times more units sold in CA compared to KY. However, there are 1,284,193 Medicaid enrollees in KY, versus 13,096,861 in California. If we normalize the number of units sold in KY, versus CA, we find that the normalized there were close to 3 times more HARVONI prescription in KY  than in CA. This is ___perhaps___ justified by the fact the KY has one of the highest rates of reported cases of Hepatitis C in the US (2.7% in KY versus 0.2% in CA).
  
https://www.cdc.gov/hepatitis/statistics/2015surveillance/pdfs/2015hepsurveillancerpt.pdf

* The number of enrollees per state was obtained here:

    https://www.medicaid.gov/medicaid/managed-care/enrollment/index.html
    
    
* A parsed/processed version (medicaid_enrollment.tsv) can be in data director of the assignment folder. Use `pandas` to load the medicaid_enrollmen file into DataFrame named `medicaid_enrollment`

In [25]:
medicaid_enrollment = pd.read_csv("data/medicaid_enrollment.tsv", sep='\t')
medicaid_enrollment.head()

STATE Total Medicaid Enrollees
0         Alabama                1,050,989
1          Alaska                  164,783
2  American Samoa                      NaN
3         Arizona                1,740,520
4        Arkansas                  762,166

* Modify `medicaid_enrollment` to uppercase all the column names 

  * Your answer can only use `DataFrame` or `Series` methods or properties
  * Do not hardcode the operation by uppercasing the columns yourself


In [26]:
medicaid_enrollment.columns = medicaid_enrollment.columns.str.upper()

* Note that some states/territories have missing values. Remove the missing values and save the resulting `DataFrame` as a new variable named `medicaid_enrollment_cleaned`

* Pay attention to how 'n/a' is given here!
* After cleaning, do you still have the Guam entry? If so, reconsider what missing values means in this context

In [27]:
medicaid_enrollment_cleaned = medicaid_enrollment.dropna()
medicaid_enrollment_cleaned = medicaid_enrollment_cleaned[medicaid_enrollment_cleaned['TOTAL MEDICAID ENROLLEES'].str.contains('n/a')== False]

### Converting `TOTAL MEDICAID ENROLLEE` Data Type

* Given that data on `TOTAL MEDICAID ENROLLEE` column contains commas on file (ex. 3,269,999 instead of 3269999), `pandas` has erroneously set the data type for that column as a string. We need to convert the column from string to `int` since we will be using it in an arithmetic expression during normalization

 

* Inspect the `dtype` property of "TOTAL MEDICAID ENROLLEES" column, and  make sure that the data type is `int`


In [28]:
medicaid_enrollment_cleaned['TOTAL MEDICAID ENROLLEES'] = medicaid_enrollment_cleaned['TOTAL MEDICAID ENROLLEES'].astype(str).str.replace(',', '').astype(int)
medicaid_enrollment_cleaned['TOTAL MEDICAID ENROLLEES'].dtype

dtype('int64')

### Associating `medicaid_data` and `medicaid_enrollment_cleaned`

- We can use the shared State information across both tables to associate both tables (SQL JOIN).
- However,  `medicaid_data` contains two-letter state abbreviations, while `medicaid_enrollment_cleaned` contains the complete state name
  - We need to convert (or append) two-letter state abbreviations to `medicaid_enrollment_cleaned`

- Pandas can read HTML and parse the code for tables. We will use that functionality to read in the state abbreviations from a Wikipedia page.
  - A brief description of what the code does is included in the comments

In [29]:
import requests

url = 'https://www.50states.com/abbreviations.htm'
header = {
    "User-Agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.75 Safari/537.36",
    "X-Requested-With": "XMLHttpRequest"
}

r = requests.get(url, headers=header)

tables = pd.read_html(r.text)


# We access the desired table by giving it's index.
# Since the URL contain only one table, then we can access that table using index 0
Codes_abbreviations = tables[0]
Codes_abbreviations.head(5)

US STATE POSTAL ABBREVIATION STANDARD ABBREVIATION
0     Alabama                  AL                  Ala.
1      Alaska                  AK                Alaska
2     Arizona                  AZ                 Ariz.
3    Arkansas                  AR                  Ark.
4  California                  CA                Calif.

* Change the the DataFrame's headers from ['US STATE', 'POSTAL ABBREVIATION', 'STANDARD ABBREVIATION'] to ['US STATE', 'ABBREVIATION', STD ABBREVIATION]

  * You can hard code this operation


In [30]:
Codes_abbreviations = Codes_abbreviations.rename({'POSTAL ABBREVIATION':'ABBREVIATION', 'STANDARD ABBREVIATION':'STD ABBREVIATION'}, axis=1)

* Combine the tables `medicaid_enrollment_cleaned` and `Codes_abbreviations` such that the resulting `DataFrame` contains all the columns in `medicaid_enrollment_cleaned` and only `ABBREVIATION` from `Codes_abbreviations` 
* Save the results to variable named `medicaid_enrollment_cleaned_with_zip`

- `medicaid_enrollment_cleaned_with_zip` should look like the following ( '...' represents remaining data that is not shown):


```
   STATE    Total Medicaid Enrollees    ABBREVIATION
0    Alabama    1,050,989    AL
1    Alaska    164,783    AK
2    Arizona    1,740,520    AZ
3    Arkansas    762,166    AR
4    California    13,096,861    CA
...
```

In [31]:
Codes_abbreviations = Codes_abbreviations.rename({'US STATE':'STATE'}, axis=1)
medicaid_enrollment_cleaned_with_zip = pd.merge(medicaid_enrollment_cleaned, Codes_abbreviations[['STATE', 'ABBREVIATION']], on='STATE')
medicaid_enrollment_cleaned_with_zip.head()

STATE  TOTAL MEDICAID ENROLLEES ABBREVIATION
0     Alabama                   1050989           AL
1      Alaska                    164783           AK
2     Arizona                   1740520           AZ
3    Arkansas                    762166           AR
4  California                  13096861           CA

* We have no further use for the column STATE in  `medicaid_enrollment_cleaned_with_zip`
  * Remove the column make sure your data in `medicaid_enrollment_cleaned_with_zip` looks like the following  ( `...` represents remaining data that is not shown):

```
    Total Medicaid Enrollees    ABBREVIATION
0   1,050,989                  AL
1   164,783                    AK
2   1,740,520                  AZ
3   762,166                    AR
4   13,096,861                 CA
....
```

In [32]:
medicaid_enrollment_cleaned_with_zip = medicaid_enrollment_cleaned_with_zip.drop(columns=['STATE'])

- Use `DataFrame medicaid_enrollment_cleaned_with_zip` to assign the appropriate number of Medicaid enrollees to each entry in the `medicaid_data`

   I.E., instead of the 10 original columns, `medicaid_data` will now have an 11th column representing the `TOTAL MEDICAID ENROLLEES` according to the STATE value in the entry.
   
- Save the resulting DataFrame into a new variable called `medicaid_data_w_enrollments`
- The resulting DataFrame should look like the following (`...` represents remaining data that is not shown):

```
UTILIZATION TYPE    STATE    NDC    PRODUCT NAME    UNITS REIMBURSED    NUMBER OF PRESCRIPTIONS    TOTAL AMOUNT REIMBURSED    MEDICAID AMOUNT REIMBURSED    NON MEDICAID AMOUNT REIMBURSED    LOCATION
0    MCOU    PA    55150023930    Dexamethas    33.0    19.0    234.98    234.98    0.0    (40.5773, -77.264)
1    FFSU    NY    23917710    ALPHAGAN P    570.0    57.0    16006.34    16006.34    0.0    (42.1497, -74.9384)
2    MCOU    OR    13925050501    Dapsone 10    456.0    15.0    1052.42    1052.42    0.0    (44.5672, -122.1269)
...
```

* The order of the columns in the `DataFrame` is not important. This answer uses the same approach as the one used to `merge` the tables above. 

In [33]:
medicaid_enrollment_cleaned_with_zip = medicaid_enrollment_cleaned_with_zip.rename({'ABBREVIATION':'STATE'}, axis=1)
medicaid_data_w_enrollments = pd.merge(medicaid_data, medicaid_enrollment_cleaned_with_zip[['STATE', 'TOTAL MEDICAID ENROLLEES']], on='STATE')

- Remove any lines where "STATE" or "PRODUCT NAME" are missing from  `medicaid_data_w_enrollments`

In [34]:
medicaid_data_w_enrollments = medicaid_data_w_enrollments.dropna( how='any', subset=['STATE', 'PRODUCT NAME'])

- Use ["STATE", "PRODUCT NAME"] as hierarchical index for `medicaid_data_w_enrollments`. Recall that a hierarchical index is simply an index with multiple levels of indexing (multiple columns)
  * Hint: the function to set an index on a `DataFrame` can take a single column name or a list of column names. The list here is  ["STATE", "NDC"]
- Call the new data `medicaid_data_w_enrollments_hierarch`
- Inspect your data to make sure the new index has now two levels (STATE and NDC)

In [35]:
medicaid_data_w_enrollments_hierach = medicaid_data_w_enrollments.set_index(['STATE', 'NDC'])
medicaid_data_w_enrollments_hierach.head(3)

UTILIZATION TYPE PRODUCT NAME  UNITS REIMBURSED  \
STATE NDC                                                           
PA    55150023930             MCOU   Dexamethas              33.0   
      456060010               FFSU   TEFLARO 60             224.0   
      115692201               MCOU   DIVALPROEX            1626.0   

                   NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
STATE NDC                                                             
PA    55150023930                     19.0                   234.98   
      456060010                       36.0                 35752.31   
      115692201                       20.0                  2318.55   

                   MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
STATE NDC                                                                       
PA    55150023930                      234.98                            0.00   
      456060010                      35752.31                            0.00   
      115692201                       2287.80                           30.75   

                             LOCATION  TOTAL MEDICAID ENROLLEES  
STATE NDC                                                        
PA    55150023930  (40.5773, -77.264)                   2569232  
      456060010    (40.5773, -77.264)                   2569232  
      115692201    (40.5773, -77.264)                   2569232


* Write a single Pandas expression to print all the lines with containing NDC 61958180101 in "PA"

 * Use a single indexing call (bracket notation) using `loc`
 
 * Hint 1: Since your index is hierarchical, `loc` is expecting two values, the first for STATE and the second for NDC


In [36]:
medicaid_data_w_enrollments_hierach.loc[('PA', 61958180101)]

/var/folders/94/dfws3bs526q9crvbr4m0tss80000gn/T/ipykernel_18718/996355958.py:1: PerformanceWarning: indexing past lexsort depth may impact performance.
  medicaid_data_w_enrollments_hierach.loc[('PA', 61958180101)]


UTILIZATION TYPE PRODUCT NAME  UNITS REIMBURSED  \
STATE NDC                                                           
PA    61958180101             FFSU   Harvoni  (             924.0   
      61958180101             MCOU   Harvoni  (            2604.0   
      61958180101             FFSU   Harvoni  (            1008.0   
      61958180101             MCOU   Harvoni  (            1932.0   
      61958180101             FFSU   Harvoni  (             924.0   
      61958180101             MCOU   Harvoni  (            3220.0   

                   NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
STATE NDC                                                             
PA    61958180101                     33.0               1017644.54   
      61958180101                     93.0               2952891.16   
      61958180101                     36.0               1107725.22   
      61958180101                     69.0               2178602.30   
      61958180101                     33.0               1015383.60   
      61958180101                    115.0               3599192.35   

                   MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
STATE NDC                                                                       
PA    61958180101                   985269.84                        32374.70   
      61958180101                  2860100.54                        92790.62   
      61958180101                  1107725.22                            0.00   
      61958180101                  2118612.30                        59990.00   
      61958180101                  1015383.60                            0.00   
      61958180101                  3482188.68                       117003.67   

                             LOCATION  TOTAL MEDICAID ENROLLEES  
STATE NDC                                                        
PA    61958180101  (40.5773, -77.264)                   2569232  
      61958180101  (40.5773, -77.264)                   2569232  
      61958180101  (40.5773, -77.264)                   2569232  
      61958180101  (40.5773, -77.264)                   2569232  
      61958180101  (40.5773, -77.264)                   2569232  
      61958180101  (40.5773, -77.264)                   2569232

#### Computing the normalized `UNITS REIMBURSED` per druc 

* Compute the `UNITS REIMBURSED` for each "NDC" in each state normalized by the number of enrollees.

- For instance in `PA`, the total `UNITS REIMBURSED` for the HARVONI `NDC` (61958180101) is 10,612

```python
total_amount_reimbursed = medicaid_data_w_enrollments_hierarch.loc[("PA", 61958180101), "UNITS REIMBURSED"].sum() 
```

- And the numebr of Medicaid Enrollees in "PA" is 2,569,232


```python
total_enrollees_PA = medicaid_data_w_enrollments_hierarch.loc["PA", "TOTAL MEDICAID ENROLLEES"].unique()[0]
```
- Therefore, the UNITS REIMBURSED per enrollee  for "HARVONI" is  0.00413041718303


```python
print(total_amount_reimbursed/total_enrollees_AK)
```

- Rather than work directly with the ratios, we are going to compute and work with thier logarithm (np.log2) instead.

  - The reason we use logs here is to avoid working small numbers. More on log-transformation in future lectures

- Save the result in a `Series` called `medicaid_reimbursement_per_enrollee`


- Your final result should be a Series that look like the following ( `...` represents data that is not shown here):
  
```
STATE  NDC    
AK     2143380    -9.609109
       2143480   -10.008280
       2322730    -6.109830
       2322830    -4.444321
       2322930    -3.855995
...

AL     2143380    -9.940595
       2143480    -9.805485
       2322730    -5.336260
...

MA     2143380    -7.921997
       2143480    -7.803463
       2144511   -13.472194
       2197590    -7.741402
       2322730    -5.414724
       2322830    -4.592154
       2322930    -4.205626

...


```


In [37]:
# Calculate the total UNITS REIMBURSED per NDC per state
total_units_reimbursed = medicaid_data_w_enrollments_hierach.groupby(['STATE', 'NDC'])['UNITS REIMBURSED'].sum()
total_enrollees = medicaid_data_w_enrollments_hierach.groupby('STATE')['TOTAL MEDICAID ENROLLEES'].unique().apply(lambda x: x[0])

# Calculate the UNITS REIMBURSED per enrollee for each NDC in each state
units_reimbursed_per_enrollee = total_units_reimbursed / total_enrollees
log_units_reimbursed_per_enrollee = np.log2(units_reimbursed_per_enrollee)
medicaid_reimbursement_per_enrollee = log_units_reimbursed_per_enrollee

print(medicaid_reimbursement_per_enrollee)

STATE  NDC        
AK     2143380        -9.609109
       2143480       -10.008280
       2322730        -6.109830
       2322830        -4.444321
       2322930        -3.855995
                        ...    
WY     76329336901    -9.773833
       76385010301    -6.146779
       76385010401    -4.707744
       76385010501    -5.965123
       76439030910    -4.794145
Length: 468370, dtype: float64


- To facilitate working with the final data, we are going to unstack `medicaid_reimbursement_per_enrollee` into a variable called  `medicaid_norm_ndc`

- Using `medicaid_reimbursement_per_enrollee`, generate a `DataFrame` where: 
  - index should be the two-letter state symbol 
  - the column names should be the NDC codes 

- The `DataFrame`  should be formatted as in the image below
  - Hint, simply unstack the data


<img src="media/unstacked.png" alt="drawing" style="width:900px;"/>


In [47]:
medicaid_reimbursement_per_enrollee = medicaid_reimbursement_per_enrollee.unstack()
medicaid_reimbursement_per_enrollee.head()

STATE           AK        AL         AR         AZ         CA         CO  \
NDC                                                                        
2143380  -9.609109 -9.940595 -12.985157 -11.533870 -10.926863 -10.862982   
2143480 -10.008280 -9.805485 -12.657103 -10.821194 -10.698372  -9.925954   
2144511        NaN       NaN        NaN        NaN -16.127018        NaN   
2144527        NaN       NaN        NaN        NaN        NaN        NaN   
2197590        NaN       NaN        NaN        NaN  -9.464364  -7.608472   

STATE           CT         DE         FL         GA  ...        SD         TN  \
NDC                                                  ...                        
2143380  -8.405103        NaN -13.534299 -15.018034  ... -9.568199 -12.240261   
2143480  -7.948475 -10.274536 -12.569558 -13.858835  ... -9.218392 -11.699134   
2144511 -12.411014        NaN -18.053374        NaN  ...       NaN        NaN   
2144527        NaN        NaN        NaN        NaN  ...       NaN        NaN   
2197590  -6.677343        NaN -10.239592        NaN  ...       NaN        NaN   

STATE           TX         UT        VA         VT         WA         WI  \
NDC                                                                        
2143380 -11.361813 -12.066772 -9.779228 -11.085710 -11.374062 -14.299344   
2143480 -10.857851 -10.625645 -9.485665 -10.407638 -11.446073 -14.884306   
2144511        NaN        NaN       NaN        NaN -14.825948        NaN   
2144527        NaN        NaN       NaN        NaN        NaN        NaN   
2197590        NaN        NaN       NaN        NaN        NaN        NaN   

STATE           WV  WY  
NDC                     
2143380 -10.526494 NaN  
2143480  -9.958528 NaN  
2144511        NaN NaN  
2144527        NaN NaN  
2197590        NaN NaN  

[5 rows x 50 columns]

#### Exploring the data (very briefly) 
- What is the drug with the highest log-normalized ratio in Hawaii?


In [54]:
# The name of the medication is Gavilyte-G.
highest_ratio_drug_hawaii = medicaid_reimbursement_per_enrollee['HI'].idxmax()
print("Drug with the highest log-normalized ratio in Hawaii:", highest_ratio_drug_hawaii)
medicaid_data_w_enrollments[medicaid_data_w_enrollments['NDC'] == 43386009019]['PRODUCT NAME'].reset_index(drop=True)[0]

Drug with the highest log-normalized ratio in Hawaii: 43386009019


'GaviLyte -'

* Investigate the NDC of the product with the highest log-normalized ratio in Hawaii 
  * What is it used for?

* Compare the value of `Units Reimbursed` that product between HI and other states, (take for instance MA, FL, OR and WA)
* Can you think for reasons why this product has the highest log-normalized ratio in Hawaii?

Gavilyte-G is a laxative that works by drawing large amounts of water into the colon.

In [55]:
# Compare the value of Units Reimbursed with other states
# MASSACHESSETTS
units_reimbursed_ma = medicaid_data_w_enrollments_hierach.groupby(['STATE', 'PRODUCT NAME'])['UNITS REIMBURSED'].sum()
pd.DataFrame(units_reimbursed_ma['MA']).loc['GAVILYTE-G']

UNITS REIMBURSED    5433.0
Name: GAVILYTE-G, dtype: float64

In [56]:
# FLORIDA
units_reimbursed_fl = medicaid_data_w_enrollments_hierach.groupby(['STATE', 'PRODUCT NAME'])['UNITS REIMBURSED'].sum()
pd.DataFrame(units_reimbursed_fl['FL']).loc['GAVILYTE-G']

UNITS REIMBURSED    1848000.0
Name: GAVILYTE-G, dtype: float64

In [57]:
# SOUTH CAROLINA
units_reimbursed_sc = medicaid_data_w_enrollments_hierach.groupby(['STATE', 'PRODUCT NAME'])['UNITS REIMBURSED'].sum()
pd.DataFrame(units_reimbursed_sc['SC']).loc['GAVILYTE-G']

UNITS REIMBURSED    1908000.0
Name: GAVILYTE-G, dtype: float64

In [58]:
# ALASKA
units_reimbursed_ak = medicaid_data_w_enrollments_hierach.groupby(['STATE', 'PRODUCT NAME'])['UNITS REIMBURSED'].sum()
pd.DataFrame(units_reimbursed_ak['AK']).loc['GAVILYTE-G']

UNITS REIMBURSED    1476000.0
Name: GAVILYTE-G, dtype: float64

* Find and list all unique `NDC`s for which the difference between the largest and second large log-normalized ratio by state is at least 10.

* For instance:
  * The highest log-normalized ratio for `00591289749` (`AZACITIDIN`) is OK where it has a log-normalized ratio  of `-1.025642`.
  * The second highest log-normalized ratio for `00591289749` is in `GA` where is has a log-normalized ration of  `-12.623428`
  


In [59]:
# Calculate the difference between the largest and second-largest log-normalized ratio by state
difference_by_state = medicaid_reimbursement_per_enrollee.apply(lambda x: x.nlargest(2).diff().iloc[-1], axis=1)
ndcs_with_difference_at_least_10 = difference_by_state[difference_by_state >= 10]

unique_ndcs = ndcs_with_difference_at_least_10.index.tolist()
print(unique_ndcs)

[]


- The Drug `AZACITIDINE` has a very high normalized UNITS REIMBURSED in OK compared to other states.
   - Normalized log value is -1.025642 (or a ratio 0.49119167009735065)
   - Second highest state has a log value of -12.623428 (0.000158478197834722)
- Oklahoma is not a high-incidence state for cancer
- Could the following explain what is happening in Oklahoma?

https://www.centerwatch.com/clinical-trials/listings/92093/acute-myeloid-leukemia-aml-study-asp2215-gilteritinib-by/?&geo_lat=35.4675602&geo_lng=-97.5164276&radius=10